File: src/java-sample/com/xtaudio/xt/CaptureDefaultInput.java
package com.xtaudio.xt;

import com.sun.jna.Pointer;
import java.io.FileOutputStream;
import java.io.IOException;

public class CaptureDefaultInput {

    static void capture(XtStream stream, Object input, Object output, int frames,
            double time, long position, boolean timeValid, long error, Object user) {

        XtFormat format = stream.getFormat();
        int sampleSize = XtAudio.getSampleAttributes(format.mix.sample).size;
        int bufferSize = frames * format.inputs * sampleSize;

        if (frames != 0)
            try {
                // Don't do this.
                ((FileOutputStream) user).write(((byte[]) input), 0, bufferSize);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
    }

    public static void main(String[] args) {

        try (XtAudio audio = new XtAudio(null, Pointer.NULL, null, null)) {
            XtService service = XtAudio.getServiceBySetup(XtSetup.CONSUMER_AUDIO);
            try (XtDevice device = service.openDefaultDevice(false)) {

                if (device == null) {
                    System.out.println("No default device found.");
                    return;
                }

                XtFormat format = new XtFormat(new XtMix(44100, XtSample.INT24), 2, 0, 0, 0);
                if (!device.supportsFormat(format)) {
                    System.out.println("Format not supported.");
                    return;
                }

                XtBuffer buffer = device.getBuffer(format);
                try (FileOutputStream recording = new FileOutputStream("xt-audio.raw");
                        XtStream stream = device.openStream(format, true, false, buffer.current, CaptureDefaultInput::capture, recording)) {
                    stream.start();
                    Thread.sleep(1000);
                    stream.stop();
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
            }
        }
    }
}

File: src/cli-sample/CaptureDefaultInput.cs
using System;
using System.IO;
using System.Threading;

namespace Xt {

    public static class CaptureDefaultInput {

        internal static void Capture(XtStream stream, Array input, Array output, int frames,
                double time, ulong position, bool timeValid, ulong error, object user) {

            XtFormat format = stream.GetFormat();
            int sampleSize = XtAudio.GetSampleAttributes(format.mix.sample).size;
            int bufferSize = frames * format.inputs * sampleSize;

            if (frames > 0)
                // Don't do this.
                ((FileStream)user).Write(((byte[])input), 0, bufferSize);
        }

        [STAThread]
        public static void Main(string[] args) {

            using (XtAudio audio = new XtAudio(null, IntPtr.Zero, null, null)) {
                XtService service = XtAudio.GetServiceBySetup(XtSetup.ConsumerAudio);
                using (XtDevice device = service.OpenDefaultDevice(false)) {

                    if (device == null) {
                        Console.WriteLine("No default device found.");
                        return;
                    }

                    XtFormat format = new XtFormat(new XtMix(44100, XtSample.Int24), 2, 0, 0, 0);
                    if (!device.SupportsFormat(format)) {
                        Console.WriteLine("Format not supported.");
                        return;
                    }

                    XtBuffer buffer = device.GetBuffer(format);
                    using (FileStream recording = new FileStream("xt-audio.raw", FileMode.Create, FileAccess.Write))
                    using (XtStream stream = device.OpenStream(format, buffer.current, new XtStreamCallback(Capture), recording)) {
                        stream.Start();
                        Thread.Sleep(1000);
                        stream.Stop();
                    }
                }
            }
        }
    }
}